Adding Token Refresh Xet Tests

rajatarya opened the pull request:
Two e2e tests that test refresh token scenario.

1. does normal upload/download with shimmed `xet_get` call to verify that
   token is refreshed in download path.

2. Directly calling `hfxet.download_files()` with `token_refresher` callback
   to ensure that hfxet calls the token refresher as expected.

Also, added download verification for `upload_large_folder` (using `snapshot_download`) for more end-to-end coverage.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2932). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

hanouticelina commented on .github/workflows/python-tests.yml:
```diff
@@ -33,6 +33,8 @@ jobs:
         include:
           - python-version: "3.13" # LFS not ran on 3.8
             test_name: "lfs"
+          - python-version: "3.11"
```
is xet not supported on python 3.8? it should be the case as we already run xet tests on this python version.

rajatarya replied:
Not sure how this got in there, removing.

hanouticelina commented on .github/workflows/python-tests.yml:
```diff
@@ -135,6 +141,9 @@ jobs:
               eval "RUN_GIT_LFS_TESTS=1 $PYTEST ../tests -k 'HfLargefilesTest'"
             ;;
 
+            xet)
```
this is a duplicate, we already isolated xet tests here: https://github.com/huggingface/huggingface_hub/blob/e4e2a049e469167f1f29f33280ebf10087f6f9ce/.github/workflows/python-tests.yml#L99

rajatarya replied:
Good catch, removing.

hanouticelina commented on src/huggingface_hub/utils/_runtime.py:
```diff
@@ -116,6 +116,15 @@ def get_fastapi_version() -> str:
     return _get_version("fastapi")
 
 
+# HF Xet
+def is_hf_xet_available() -> bool:
+    return is_package_available("hf_xet")
+
+
+def get_hf_xet_version() -> str:
+    return _get_version("hf_xet")
+
+
```
this is a duplicate, it's already implemented here: https://github.com/huggingface/huggingface_hub/blob/de81c60411ccfdf38112b51dca05236b663182c7/src/huggingface_hub/utils/_runtime.py#L165

hanouticelina commented on tests/test_xet_integration.py:
```diff
@@ -0,0 +1,325 @@
+
+
+@require_hf_xet
+class TestHfXet(TestCase):
```
I think we can move `test_force_refresh_token` and `test_hfxet_direct_download_files_token_refresher` to [tests/test_xet_download.py](https://github.com/huggingface/huggingface_hub/blob/xet-integration/tests/test_xet_download.py)

Also, I just realized that `test__xet_available`, `test_upload_and_download_with_hf_xet`, `test_upload_folder_download_snapshot_with_hf_xet` and `test_upload_large_folder_download_snapshot_with_hf_xet` might be unnecessary, we already have some e2e flow tested in [tests/test_xet_upload.py](https://github.com/huggingface/huggingface_hub/blob/xet-integration/tests/test_xet_upload.py). 

My suggestion would be to enhance the existing tests in `test_xet_upload.py` and/or `test_xet_download.py` to cover any missing scenarios, wdyt?

hanouticelina reviewed (commented):
thanks @rajatarya for taking care of this! I left a couple of comments, let me know what you think 🙂

rajatarya commented on tests/test_xet_integration.py:
```diff
@@ -0,0 +1,325 @@
+        _p = PyPointerFile("path", "hash", 100)
+    except ImportError:
+        return False
+    return True
```
Removed test_xet_integration.py.

rajatarya commented on tests/test_xet_integration.py:
```diff
@@ -0,0 +1,325 @@
+    if not is_hf_xet_available():
+        return skip("Test requires hf_xet")(test_case)
+    else:
+        return test_case
```
removed test_xet_integration.py

rajatarya commented:
I believe this PR is ready for final review. I believe I have addressed all the feedback (thanks @hanouticelina, great feedback!)

My apologies - I should have better tracked the tests in the `test_xet_upload.py` and `test_xet_download.py` - they are better organized, better structured, cleaner, and much more thorough than the tests I wrote. And as those tests were merged they had already set up the GitHub Action workflow changes for xet so this PR was duplicating (poorly) a bunch of that.

Now this PR is simply adding two token refresh tests on to `test_xet_download.py`, and adding download + byte verification of `upload_large_folder` using `snapshot_download` in `test_xet_upload.py`.

Wauplin commented on tests/test_xet_upload.py:
```diff
@@ -240,3 +240,15 @@ def test_upload_large_folder(self, api, tmp_path, repo_url: RepoUrl) -> None:
+            for j in range(N_FILES_PER_FOLDER):
+                for filename in [f"subfolder_{i}/file_xet_{i}_{j}.bin", f"subfolder_{i}/file_regular_{i}_{j}.txt"]:
+                    local_file = Path(folder) / filename
+                    downloaded_file = hf_hub_download(repo_id=repo_id, filename=filename, cache_dir=None)
```
Just to be sure to understand, is this test supposed to check the downloaded content is the same as the uploaded one? or is it to check that both "download to cache" and "download to local_dir" are working and retrieving the same content? 

To check that download works correctly, I'd just do the `snapshot_download` as you did and then iterate on `i/j`  and check the content is `f"content_lfs_{i}_{j}"`/`f"content_regular_{i}_{j}"`

rajatarya replied:
The intent is to verify that the bytes are downloaded (no cache used) and contain the same contents as they exist locally. I'll change it to check the contents directly instead of comparing to the local files.

Wauplin reviewed (approved):
Thanks for iterating and for the detailed tests! LGTM except the last test where I'm unsure what is supposed to be tested. Also let's wait for @hanouticelina 's approval first

hanouticelina commented on tests/test_xet_upload.py:
```diff
@@ -240,3 +250,153 @@ def test_upload_large_folder(self, api, tmp_path, repo_url: RepoUrl) -> None:
+
+@requires("hf_xet")
+class TestXetE2E(TestXetUpload):
+    def test_force_refresh_token(self, api, repo_url, tmp_path):
```
I'm not sure I understand how this tests the token refresh process, I don't see a mock of the `token_refresher` function as it's done in `test_hf_xet_with_token_refresher`. is it by checking that `xet_get` is called?

rajatarya replied:
This test simply changes the expiration time of the token to now(), ensuring the refresh should happen immediately. However, by not mocking the token_refresh function itself it doesn't add anything to the overall test coverage.

I'll drop this test, I don't think it is adding anything over the `test_hf_xet_with_token_refresher` test.

rajatarya replied:
The only thing this test covers that isn't covered by `test_hf_xet_with_token_refresher` is that it starts from the user function `hf_hub_download` so is more of a an end-to-end test.

`test_hf_xet_with_token_refresher` is more of a lower-level regression test that ensures `hf_xet` doesn't change its interface. It calls `download_files` directly, which isn't something a user can do. 

That is why I originally had both tests, but I don't think it is worth maintaining both tests given how subtle the coverage point is.

rajatarya commented:
OK now simplified to only one new test for token refresh. Thanks for all the great feedback!

rajatarya commented:
> OK now simplified to only one new test for token refresh. Thanks for all the great feedback!

Am I good to merge? @bpronan @hanouticelina can you approve if you think ready for merging?

hanouticelina reviewed (approved):
Looks good, thank you @rajatarya! I pushed [5310054](https://github.com/huggingface/huggingface_hub/pull/2932/commits/5310054da9d83f5a39da489524927028da4373d9) to simplify a bit the `test_upload_large_folder` test to just call `snapshot_download` and then check that the content matches.
